### ISSUE TITLE
Pyinstaller for Salt

### ISSUE THREAD

**Issue by kravivar** [pyinstaller/pyinstaller#2477]
Hi Team,

I`ve been trying to get the salt within pyinstaller binary for one of our project.

The implementation is very simple, just wanted to get the grains from OS.

## Problem: 

The pyinstaller is not able to identify the salt modules only on windows it is perfectly working fine on Linux distros.

## Steps to recreate:
* No need to install python or pip the below steps does everything.
* Install salt dev on any windows machine using : https://docs.saltstack.com/en/latest/topics/installation/windows.html#create-a-build-environment
* pip install pyinstaller
* Download the two files attached into directory : [Archive.zip](https://github.com/pyinstaller/pyinstaller/files/809217/Archive.zip)
* run pyinstaller `pyinstaller --hidden-import=salt --additional-hooks-dir=./ --onedir .\grains.py`

## Actual output from running python
```
PS C:\hubble-dev\hubble\back> python .\grains.py
loaded ok ...
PS C:\hubble-dev\hubble\back>
```

## Output from pyinstaller
```
PS C:\hubble-dev\hubble\back> .\dist\hubble\grains.exe
[Error 126] The specified module could not be found
loaded ok ...
PS C:\hubble-dev\hubble\back>
```

I have tried out the following docs,
https://github.com/pyinstaller/pyinstaller/wiki/If-Things-Go-Wrong

Any guidance is highly appreciated.

Thanks,
Kripa

**Comment by htgoebel** [comment]
What does "get the grains" mean?

Did you already compare the build results between GNU/Linux and Windows? The logs created in the build directory should give all relevant information. Also archive_viewer might help.

Have you also followed*all* of [Make sure everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly)?

Please note that I will not download some zip-Archive. Please provide the files as a gist or something like this.

**Comment by kravivar** [comment]
@htgoebel : Thank you very much for your response. Yes I have tried out the debug options like dependency walker, archive viewer I found that its not able to import `salt.ext.six.moves` module which is in the warning : https://github.com/saltstack/salt/blob/develop/salt/ext/six.py#L171

The grains is salt terminology and it basically means information operating system, Example as follows.

## Sample output of grains
```
PS C:\hubble-dev\hubble\back> python .\hubble.py
{'biosversion': u'Revision: 1.221', 'kernel': 'Windows', 'domain': 'us-west-1.compute.internal', 'zmqversion': '4.1.2',
'kernelrelease': u'6.3.9600', 'motherboard': {'serialnumber': None, 'productname': None}, 'pythonpath': ['C:\\hubble-dev
\\hubble\\back', 'C:\\Python27\\python27.zip', 'C:\\Python27\\DLLs', 'C:\\Python27\\lib', 'C:\\Python27\\lib\\plat-win',
 'C:\\Python27\\lib\\lib-tk', 'C:\\Python27', 'C:\\Python27\\lib\\site-packages', 'C:\\Python27\\lib\\site-packages\\pyy
aml-3.11-py2.7-win-amd64.egg', 'C:\\Python27\\lib\\site-packages\\win32', 'C:\\Python27\\lib\\site-packages\\win32\\lib'
, 'C:\\Python27\\lib\\site-packages\\Pythonwin'], 'serialnumber': u'ec26c7d5-fcb0-06fd-c92a-d617bc0657ff', 'pid': 2904,
'ip_interfaces': {u'Intel(R) 82599 Virtual Function': [u'172.31.22.234', u'fe80::9095:98e3:3231:238d']}, 'groupname': ''
, 'shell': '/bin/sh', 'mem_total': 8191, 'gpus': [], 'timezone': u'(UTC) Coordinated Universal Time', 'id': 'WIN-EFNVGMQ
SKK5.us-west-1.compute.internal', 'osrelease': '2012ServerR2', 'ps': 'tasklist.exe', 'server_id': 1491098507L, 'ip6_inte
rfaces': {u'Intel(R) 82599 Virtual Function': [u'fe80::9095:98e3:3231:238d']}, 'num_cpus': 2, 'hwaddr_interfaces': {u'In
tel(R) 82599 Virtual Function': u'02:00:65:A6:AB:AA'}, 'osfullname': u'Microsoft Windows Server 2012 R2 Standard', 'ip4_
interfaces': {u'Intel(R) 82599 Virtual Function': [u'172.31.22.234']}, 'init': 'Windows', 'master': 'salt', 'ipv4': ['17
2.31.22.234'], 'ipv6': ['fe80::9095:98e3:3231:238d'], 'osversion': u'6.3.9600', 'localhost': 'WIN-EFNVGMQSKK5', 'virtual
_subtype': 'HVM domU', 'username': 'Administrator', 'fqdn_ip4': [], 'saltpath': 'C:\\Python27\\lib\\site-packages\\salt'
, 'fqdn_ip6': [], 'nodename': 'WIN-EFNVGMQSKK5', 'saltversion': '2016.11.0-882-g8809d33', 'saltversioninfo': [2016, 11,
0, 0], 'locale_info': {'detectedencoding': 'cp437', 'defaultlanguage': 'en_US', 'defaultencoding': 'cp1252'}, 'cpu_model
': u'Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz', 'host': 'WIN-EFNVGMQSKK5', 'os_family': 'Windows', 'osfinger': 'Windows
-2012ServerR2', 'pythonversion': [2, 7, 12, 'final', 0], 'manufacturer': u'Xen', 'num_gpus': 0, 'virtual': 'Xen', 'osser
vicepack': None, 'fqdn': 'WIN-EFNVGMQSKK5.us-west-1.compute.internal', 'pythonexecutable': 'C:\\Python27\\python.exe', '
productname': u'HVM domU', 'cpuarch': 'AMD64', 'osmanufacturer': u'Microsoft Corporation', 'osrelease_info': (2012, 2),
'path': 'C:\\Python27\\;C:\\Python27\\Scripts;C:\\Windows\\system32;C:\\Windows;C:\\Windows\\System32\\Wbem;C:\\Windows\
\System32\\WindowsPowerShell\\v1.0\\;C:\\Program Files\\Amazon\\cfn-bootstrap\\;C:\\ProgramData\\chocolatey\\bin;C:\\Pro
gram Files\\Git\\cmd;C:\\Python27\\lib\\site-packages\\pywin32_system32;/bin;/sbin;/usr/bin;/usr/sbin;/usr/local/bin', '
os': 'Windows', 'windowsdomain': u'WORKGROUP'}
PS C:\hubble-dev\hubble\back>
```


## Grains.py script
```
grains.py

#!/usr/bin/env python

try:
    import salt
    import salt.config
    import salt.loader

#    __opts__ = salt.config.minion_config('C:\hubble\hubble.txt')
#    __grains__ = salt.loader.grains(__opts__)
except Exception, e:
    print e

print "loaded ok ..."
#print __grains__
```

## hook-salt.py
```
# -*- coding: utf-8 -*-
'''
    :codeauthor: :email:`Pedro Algarvio (pedro@algarvio.me)`
    :copyright: © 2016 by the SaltStack Team, see AUTHORS for more details.
    :license: Apache 2.0, see LICENSE for more details.


    hook-salt.py
    ~~~~~~~~~~~~

    @todo: add description
'''
import os
from PyInstaller.utils.hooks import (collect_data_files,
                                     collect_submodules,
                                     collect_dynamic_libs,
                                     PY_EXECUTABLE_SUFFIXES)

DATAS = []
BINARIES = []
HIDDEN_IMPORTS = []
# Let's handle salt.
# For salt we include the loader modules source files in DATAS and remove those from HIDDEN_IMPORTS
# This is the only way the salt loader seems to work
HIDDEN_IMPORTS.extend(collect_submodules('salt'))
DATAS.extend(collect_data_files('salt', subdir=".", include_py_files=True))

# Let's filter out salt loader modules which are included as data files
SALT_LOADERS = [
    'salt.auth',
    'salt.beacons',
    'salt.cloud.clouds',
    'salt.engines',
    'salt.executors',
    'salt.grains',
    'salt.log.handlers',
    'salt.modules',
    'salt.netapi',
    'salt.output',
    'salt.output',
    'salt.pillars',
    'salt.proxy',
    'salt.queues',
    'salt.renderers',
    'salt.returners',
    'salt.roster',
    'salt.runners',
    'salt.sdb',
    'salt.search',
    'salt.serializers',
    'salt.spm.pkgdb',
    'salt.spm.pkgfiles',
    'salt.states',
    'salt.thorium',
    'salt.tops',
    'salt.utils',
    'salt.wheels',
]
LOADER_MODULES_SOURCES = []
for sloader in SALT_LOADERS:
    for mod in HIDDEN_IMPORTS[:]:
        if mod == sloader:
            continue
        if mod.startswith(mod):
            LOADER_MODULES_SOURCES.append(mod.replace('.', os.sep))

# Let's remove any python source files included that are in HIDDEN_IMPORTS but not on DATAS
for entry in DATAS[:]:
    path, mod = entry
    if not path.endswith(tuple(PY_EXECUTABLE_SUFFIXES)):
        # We are only after python files
        continue
    no_ext_path = os.path.splitext(path)[0]
    if not no_ext_path.endswith(tuple(LOADER_MODULES_SOURCES)):
        if entry in DATAS:
            DATAS.remove(entry)

# Some packages salt required, which we should include that are not discovered by PyInstaller
PACKAGES = []

for pkg in PACKAGES:
    DATAS.extend(collect_data_files(pkg, include_py_files=True))
    BINARIES.extend(collect_dynamic_libs(pkg))
    HIDDEN_IMPORTS.extend(collect_submodules(pkg))

# Finally, define the globals that PyInstaller expects
hiddenimports = HIDDEN_IMPORTS
datas = DATAS
binaries = BINARIES

```

**Comment by htgoebel** [comment]
Thanks for your reply.

> import salt.ext.six.moves module which is in the

So salt is yet another package bringing it's own copy of six. I added a pre-_save-import-hook for handling this. Please check out branch `wip-salt-ext-moves` and test it. If you can make the `test_salt_ext_six` in  `tests/functional/test_libraries.py` work (which means: fial if the hook is not present), this would be fine.

Regarding the hook you posted above: It looks, well, uff, ... ugly. I've never seen such a complicated hook. Instead of including dedicated modules *if* (and only if) they are required, it pulls in many modules, while using another long black-list (which contains typos, too). Also is searches many packages for data-files, while salt (as far as I can see) only has data-files in templates, daemons/flo, and cloud/deploy.

Where is this hook coming from?

**Comment by kravivar** [comment]
@htgoebel The hook was given to me by one of my colleague : https://gitlab.com/basepi/salt-ci-builder/blob/master/.ci/pyinstaller/hooks/hook-salt.py

Not sure if you have access to that, but If you have gitlab account I think I can ask him to grant you access. 

I have tested out the result and tests pass,

```
PS C:\hubble-dev\pyinstaller\tests> py.test tests/functional/test_libraries.py -k test_salt_ext_six
============================= test session starts =============================
platform win32 -- Python 2.7.12, pytest-3.0.6, py-1.4.32, pluggy-0.4.0 -- c:\python27\python.exe
cachedir: ..\.cache
rootdir: C:\hubble-dev\pyinstaller, inifile: setup.cfg
plugins: catchlog-1.2.2, drop-dup-tests-0.1.0, timeout-1.2.0, xdist-1.15.0

======================== no tests ran in 0.01 seconds =========================
ERROR: file not found: tests/functional/test_libraries.py
PS C:\hubble-dev\pyinstaller\tests>
PS C:\hubble-dev\pyinstaller\tests>
PS C:\hubble-dev\pyinstaller\tests> py.test functional/test_libraries.py -k test_salt_ext_six
============================= test session starts =============================
platform win32 -- Python 2.7.12, pytest-3.0.6, py-1.4.32, pluggy-0.4.0 -- c:\python27\python.exe
cachedir: ..\.cache
rootdir: C:\hubble-dev\pyinstaller, inifile: setup.cfg
plugins: catchlog-1.2.2, drop-dup-tests-0.1.0, timeout-1.2.0, xdist-1.15.0
collected 103 items

functional\test_libraries.py::test_salt_ext_six[onedir] PASSED
functional\test_libraries.py::test_salt_ext_six[onefile] PASSED

============================ 101 tests deselected =============================
================== 2 passed, 101 deselected in 46.92 seconds ==================```

The story behind that is that pyinstaller packaging the salt using a hidden-import doesnt work ... I havent really investigated why it is so, now I`m thinking about it. May be that could answer questions ...

**Comment by htgoebel** [comment]
```
  File "site-packages\PyInstaller\loader\pyiboot01_bootstrap.py", line 136, in __init__
  File "ctypes\__init__.py", line 362, in __init__
WindowsError: [Error 126] The specified module could not be found
[Error 126] The specified module could not be found
loaded ok ...
PS C:\hubble-dev\hubble\back> python
Python 2.7.12 (v2.7.12:d33e0cf91556, Jun 27 2016, 15:24:40) [MSC v.1500 64 bit (AMD64)] on win32```

Argl! Didn't you say , you followed *all* of the steps in It Thinks go wrong? Obviously you missed running with `--debug`. Thanks for wasting my time :-\

This bug is *not* related to salt, but to some dll not found in windows. See https://github.com/saltstack/salt/blob/develop/salt/utils/rsax931.py#L30. You should see a about warning in the output of PyInstaller.

I'm closing this issue now, since it is full of information leading into the wrong direction. If you still have problems, please open a new issue.

**Comment by kravivar** [comment]
@htgoebel : Thank you for helping me out with the issue.

the issue was resolved by adding an explicit binary for the missing lib,

```
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          a.binaries + [('libeay32.dll', 'c:\\windows\\system32\\libeay32.dll', 'BINARY')],
          name='bitmessagemain',
          debug=False,
          strip=None,
          upx=True,
          console=False , icon='src\\images\\can-icon.ico')
```
 
Apologies for any inconvenience caused.
